Tools/Scripts/webkitpy/w3c/test_importer.py

```diff
@@ -209,7 +209,12 @@ def __init__(self, port, test_paths, options):
         }
 
     def do_import(self):
-        if not self.source_directory:
+        if self.source_directory:
+            # We need to convert the path as a string to make the tests framework happy
+            source_path = str(Path(self.source_directory) / 'web-platform-tests')
+            git = self.test_downloader().git(source_path)
```

Comment:
hmm so I guess my two remaining failues is because I bring the full test_downloader in a scenario which was not expected for the tests cases before. 

```
OpenSource % /usr/bin/python3 ./Tools/Scripts/test-webkitpy               
Suppressing most webkitpy logging while running unit tests.
Skipping tests in the following modules or packages because they are really, really, slow:
    webkitpy.common.checkout.scm.scm_unittest
    (https://bugs.webkit.org/show_bug.cgi?id=31818; use --all to include)

[1/2630] reporelaypy.tests.checkoutroute_unittest.RedirectorUnittest.test_fallback passedMock process 44483 send signal 9
Mock process 44496 send signal 9
Mock process 44496 send signal 9
[25/2630] webkitbugspy.tests.bugzilla_unittest.TestBugzilla.test_comments passedMock process 44484 send signal 9
Mock process 44488 send signal 9
[393/2630] webkitpy.common.net.bugzilla.bugzilla_unittest.BugzillaQueriesTest.test_load_query passed/Users/karlcow/code/OpenSource/Tools/Scripts/libraries/autoinstalled/python-3-arm64/bs4/builder/__init__.py:545: XMLParsedAsHTMLWarning: It looks like you're parsing an XML document using an HTML parser. If this really is an HTML document (maybe it's XHTML?), you can ignore or filter this warning. If it's XML, you should know that using an XML parser will be more reliable. To parse this document as XML, make sure you have the lxml package installed, and pass the keyword argument `features="xml"` into the BeautifulSoup constructor.
  warnings.warn(
[874/2630] webkitpy.layout_tests.run_webkit_tests_integrationtest.RunTest.test_single_file passedProcess Process-8:4:                                       
[1483/2630] webkitpy.layout_tests.run_webkit_tests_integrationtest.RunTest.test_skipped_flag (+9)worker/1 starting
worker/0 starting
worker/0 stopping
worker/1 stopping
[1776/2630] webkitpy.w3c.test_importer_unittest.TestImporterTest.test_import_dir_with_no_tests_and_no_hg erred:                           
  Traceback (most recent call last):
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/common/memoized.py", line 42, in __call__
      return self._results_cache[args]
  KeyError: (<webkitpy.port.config.Config object at 0x10c071fa0>,)
  
  During handling of the above exception, another exception occurred:
  
  Traceback (most recent call last):
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/w3c/test_importer_unittest.py", line 85, in test_import_dir_with_no_tests_and_no_hg
      importer.do_import()
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/w3c/test_importer.py", line 215, in do_import
      git = self.test_downloader().git(source_path)
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/w3c/test_downloader.py", line 102, in git
      return Git(test_repository, None, executive=self._host.executive, filesystem=self._filesystem)
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/common/checkout/scm/git.py", line 65, in __init__
      SCM.__init__(self, cwd, **kwargs)
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/common/checkout/scm/scm.py", line 68, in __init__
      self.checkout_root = self.find_checkout_root(self.cwd)
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/common/checkout/scm/git.py", line 131, in find_checkout_root
      checkout_root = self._run_git(['rev-parse', '--show-toplevel'], cwd=(path or "./")).strip()
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/common/checkout/scm/git.py", line 110, in _run_git
      return self.run(full_command_args, **full_kwargs)
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/common/checkout/scm/scm.py", line 73, in run
      return self._executive.run_command(args,
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/common/system/executive_mock.py", line 230, in run_command
      raise self._exception  # pylint: disable=E0702
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/port/config.py", line 123, in _read_configuration
      configuration_path = self._filesystem.join(self.build_directory(None), "Configuration")
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/port/config.py", line 78, in build_directory
      output = self._executive.run_command(args, cwd=self._webkit_finder.webkit_base(), return_stderr=False).rstrip()
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/common/system/executive_mock.py", line 230, in run_command
      raise self._exception  # pylint: disable=E0702
  OSError
  
[2103/2630] webkitscmpy.test.git_unittest.TestGit.test_files_changed passedMock process 44493 send signal 9                               
[2140/2630] webkitpy.w3c.test_importer_unittest.TestImporterTest.test_import_dir_with_no_tests erred:
  Traceback (most recent call last):
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/common/memoized.py", line 42, in __call__
      return self._results_cache[args]
  KeyError: (<webkitpy.port.config.Config object at 0x134252550>,)
  
  During handling of the above exception, another exception occurred:
  
  Traceback (most recent call last):
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/w3c/test_importer_unittest.py", line 97, in test_import_dir_with_no_tests
      importer.do_import()
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/w3c/test_importer.py", line 215, in do_import
      git = self.test_downloader().git(source_path)
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/w3c/test_downloader.py", line 102, in git
      return Git(test_repository, None, executive=self._host.executive, filesystem=self._filesystem)
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/common/checkout/scm/git.py", line 65, in __init__
      SCM.__init__(self, cwd, **kwargs)
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/common/checkout/scm/scm.py", line 68, in __init__
      self.checkout_root = self.find_checkout_root(self.cwd)
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/common/checkout/scm/git.py", line 131, in find_checkout_root
      checkout_root = self._run_git(['rev-parse', '--show-toplevel'], cwd=(path or "./")).strip()
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/common/checkout/scm/git.py", line 110, in _run_git
      return self.run(full_command_args, **full_kwargs)
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/common/checkout/scm/scm.py", line 73, in run
      return self._executive.run_command(args,
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/common/system/executive_mock.py", line 230, in run_command
      raise self._exception  # pylint: disable=E0702
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/port/config.py", line 123, in _read_configuration
      configuration_path = self._filesystem.join(self.build_directory(None), "Configuration")
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/port/config.py", line 78, in build_directory
      output = self._executive.run_command(args, cwd=self._webkit_finder.webkit_base(), return_stderr=False).rstrip()
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/common/system/executive_mock.py", line 230, in run_command
      raise self._exception  # pylint: disable=E0702
  webkitpy.common.system.executive.ScriptError: abort: no repository found in '/Volumes/Source/src/wk/Tools/Scripts/webkitpy/w3c' (.hg not found)!
  
Ran 2630 tests in 20.969s                                                                                                                     
FAILED (failures=0, errors=2)
```

Reply:
This test https://github.com/WebKit/WebKit/blob/45c4d855298d4aa4bdb43aefb0ba8b69af92f888/Tools/Scripts/webkitpy/w3c/test_importer_unittest.py#L74-L86

And this test
https://github.com/WebKit/WebKit/blob/45c4d855298d4aa4bdb43aefb0ba8b69af92f888/Tools/Scripts/webkitpy/w3c/test_importer_unittest.py#L87-L97

Reply:
The two tests are here to check when calling `do_import()` with empty directory and with empty directory and no checkout. 

my modification switches from a model where it was 

```diff
-        if not self.source_directory:
+        if self.source_directory:
+            …
+        else:
```

Reply:
unittests are harder to write when the function, here `do_import`, does too many things.

Reply:
ok cool. I wrapped the in a try/except clause. Because indeed if the source directory is not a checkout, it would fail.